### PR TITLE
feat: T-19102 add validate relay func

### PR DESCRIPTION
**Pull request by pablocampogo**
Add validate relay function needed to validate relays before they're sent to DB

### CODE REVIEW

File: types/relay.go [hunk 1]
```diff
@@ -18,7 +46,7 @@ type Relay struct {
 	PoktNodeAddress          string        `json:"poktNodeAddress"`
 	RelayStartDatetime       time.Time     `json:"relayStartDatetime"`
 	RelayReturnDatetime      time.Time     `json:"relayReturnDatetime"`
-	IsError                  bool          `json:"isError"`
+	IsError                  bool          `json:"isError"` // this field must be before the other error fields for validation to work
```
Comment (rem1niscence) [inline types/relay.go:49]: I don't understand the comment, as this data structure is unordered, how it could be set before?

File: types/relay.go [hunk 2]
```diff
@@ -36,3 +64,42 @@ type Relay struct {
...
+		fieldName := structType.Field(i).Name
+
+		isSet := field.IsValid() && !field.IsZero()
+
+		if isSet {
+			if fieldName == "IsError" {
+				isError = true
+			}
+
+			if shouldBeEmptyField[fieldName] || (!isError && errorField[fieldName]) {
```
Comment (rem1niscence) [inline types/relay.go:91]: going along with the previous comment, what if the errorField comes before `isError`, assuming this is an unordered slice ofc

File: types/relay.go [hunk 3]
```diff
@@ -36,3 +64,42 @@ type Relay struct {
...
+				return fmt.Errorf("%s should not be set", fieldName)
+			}
+
+			if fieldName == "ErrorType" && !validErrorTypes[field.String()] {
+				return fmt.Errorf("%s is not valid", fieldName)
+			}
+		}
+
+		if !isSet {
+			if shouldBeEmptyField[fieldName] || field.Kind() == reflect.Bool || (!isError && errorField[fieldName]) {
```
Comment (rem1niscence) [inline types/relay.go:105]: A comment here on what is checking will be helpful for the future to minimize cognitive load

**Review by rem1niscence — COMMENTED** [review]
overall looks good imo, left some comments to clarify some things

File: types/relay.go (hunk 1 above)
Comment (pablocampogo) [reply types/relay.go:49]: The field are always reviewed in the order they are on the implementation

File: types/relay.go (hunk 2 above)
Comment (pablocampogo) [reply types/relay.go:91]: It is not unordered

**Review by rem1niscence — APPROVED** [review]
lgtm=

**Pull request merged**